test/testutil.h

```diff
@@ -141,10 +143,94 @@ __owur int run_tests(const char *test_prog_name);
+    DECLARE_COMPARISON(type, name, lt)                                  \
+    DECLARE_COMPARISON(type, name, le)                                  \
+    DECLARE_COMPARISON(type, name, gt)                                  \
+    DECLARE_COMPARISON(type, name, ge)
+
+#define DEFINE_COMPARISON(type, name, opname, op, fmt)                  \
+    int name ## _ ## opname(const type t1, const type t2,               \
+                            const char *desc, ...)                      \
+    {                                                                   \
+        va_list ap;                                                     \
+        if (t1 op t2)                                                   \
+            return 1;                                                   \
+        va_start(ap, desc);                                             \
+        fail_message(desc, ap, #type, fmt " " #opname " " fmt, t1, t2); \
+        va_end(ap);                                                     \
+        return 0;                                                       \
+    }
+
+#define DEFINE_COMPARISONS(type, name, fmt)                             \
+    DEFINE_COMPARISON(type, name, eq, ==, fmt)                          \
+    DEFINE_COMPARISON(type, name, ne, !=, fmt)                          \
+    DEFINE_COMPARISON(type, name, lt, <, fmt)                           \
+    DEFINE_COMPARISON(type, name, le, <=, fmt)                          \
+    DEFINE_COMPARISON(type, name, gt, >, fmt)                           \
+    DEFINE_COMPARISON(type, name, ge, >=, fmt)
+
+#define IMPLEMENT_COMPARISONS(type, name, fmt)                          \
+    DECLARE_COMPARISONS(type, name)                                     \
+    DEFINE_COMPARISONS(type, name, fmt)
+
+/*
+ * Define eight comparisons between pairs of integers and unsigned integers.
+ * These functions return 1 if the test is true.
+ * Otherwise, they return 0 and pretty-print diagnostics.
+ */
+DECLARE_COMPARISONS(int, int)
+DECLARE_COMPARISONS(unsigned int, uint)
+
+/*
+ * Pointer comparisons against other pointers and null.
+ * These functions return 1 if the test is true.
+ * Otherwise, they return 0 and pretty-print diagnostics.
+ */
+DECLARE_COMPARISON(void *, ptr, eq)
+DECLARE_COMPARISON(void *, ptr, ne)
+int ptr_null(const void *, const char *, ...);
+int ptr_not_null(const void *, const char *, ...);
+
+/*
+ * Equality test for strings where NULL is a legitimate value.
+ * These calls return 1 if the two passed strings compare true.
+ * Otherwise, they return 0 and pretty-print diagnostics.
+ */
+DECLARE_COMPARISON(char *, strings, eq)
+DECLARE_COMPARISON(char *, strings, ne)
+
+/*
+ * Check a boolean result for being true or false.
+ * They return 1 if the condition is true.
+ * Otherwise, they return 0 and pretty-prints diagnostics using |desc|.
  */
-int strings_equal(const char *desc, const char *s1, const char *s2);
+int bool_true(int, const char *desc, ...);
```

Comment:
Ah, so you were thinking `TEST_ptr` *in addition* to `TEST_ptr_eq` and `TEST_ptr_ne`? That escaped me, and in that case I have no concerns in the matter...